gr-blocks/lib/rotator_cc_impl.cc

```diff
@@ -40,6 +48,33 @@ void rotator_cc_impl::set_phase_inc(double phase_inc)
     d_r.set_phase_incr(exp(gr_complex(0, phase_inc)));
 }
 
+void rotator_cc_impl::handle_phase_inc_msg(pmt::pmt_t msg)
+{
+    const pmt::pmt_t inc_key = pmt::intern("inc");
+    const pmt::pmt_t offset_key = pmt::intern("offset");
+
+    if (pmt::is_dict(msg)) {
+        /* New phase increment (mandatory) */
+        if (!pmt::dict_has_key(msg, inc_key))
+            throw std::runtime_error("rotator_cc: Control message "
+                                     "must contain \"inc\" key");
+
+        const double phase_inc =
+            pmt::to_double(pmt::dict_ref(msg, inc_key, pmt::PMT_NIL));
+
+        /* Absolute sample offset when update is to be applied (optional) */
+        const uint64_t update_offset =
+            pmt::dict_has_key(msg, offset_key)
+                ? pmt::to_uint64(pmt::dict_ref(msg, offset_key, pmt::PMT_NIL))
+                : nitems_written(0) + 1; // update right away when not provided
+
+        d_inc_update_queue.push(std::make_pair(update_offset, phase_inc));
```

Comment:
Almost! You  need to hold a mutex while accessing the queue, because the message handling and the work function are going to happen in separate threads. Since you're not already using it, using `d_setlock` is probably fine.

Reply:
Here's the code, for both here and in `::work`:
```
gr::thread::scoped_lock guard(d_setlock);
```
Using the overall block lock like this is the way to go.

Reply:
Add the `scoped_lock` to `handle_ctrl_msg()` and `work()` as suggested.